### DISCUSSION TITLE
How to model a list of values for one field with checkboxes?

### DISCUSSION THREAD

**Discussion by egze** [woylie/flop_phoenix#189] [Q&A]
We want to implement this filter:
<img width="297" alt="Monitoring · Orchestrator 2023-04-28 18-06-50" src="https://user-images.githubusercontent.com/29657/235198526-ce90128a-c250-4185-b07d-6c553f94411e.png">

The schema is the following:
```elixir
@derive {
    Flop.Schema,
    custom_fields: [
      status: [
        filter: {ExecutionQueries, :status_filter, [source: :status]},
        ecto_type: :string
      ]
    ]
  }
```

I want it to be an `ANY ()` query. Something like this `WHERE status ANY ('succeeded', 'running')`

I think I need to get my params in such shape:
```elixir
%{"filters" => %{"0" => %{"field" => "status", "op" => "in", "value" => ["succeeded", "running"]}}}
```

But I'm lost how to model it with checkboxes. Any ideas?

**Comment by woylie** [comment]
That's more a question for Phoenix than for Flop.

Based on the [input component](https://github.com/phoenixframework/phoenix/blob/7547df06b420608b9eae85bba732387cf3739b05/priv/templates/phx.gen.live/core_components.ex#L227) as generated by Phoenix 1.7, I like to define an additional `checkbox-group` type, and also a `radio-group` type. Both of these input types take the options in exactly the same format as the `select` type (as returned by [options_for_select/2](https://hexdocs.pm/phoenix_html/Phoenix.HTML.Form.html#options_for_select/2)), which allows you to switch between selects, checkboxes and radio buttons by just changing the `type` attribute.

Roughly, the function clause for the `checkbox-group` looks like this:

```elixir
def input(%{type: "checkbox-group"} = assigns) do
  ~H"""
  <div phx-feedback-for={@name <> "[]"}>
    <fieldset>
      <legend><%= @label %></legend>
      <div class="checkbox-group">
        <.label :for={option <- @options} class="checkbox"}>
          <input
            type="checkbox"
            id={(@id || @name) <> "_#{option_value(option)}"}
            name={@name <> "[]"}
            value={option_value(option)}
            checked={selected?(option_value(option), @value)}
            {@rest}
          />
          <span><%= option_label(option) %></span>
        </.label>
      </div>
    </fieldset>
  </div>
  """
end

defp option_label({label, _}), do: label
defp option_label(label), do: label

defp option_value({_, value}), do: value
defp option_value(value), do: value

defp selected?(option, value) when is_list(value) do
  value_list = Enum.map(value, &html_escape/1)
  html_escape(option) in value_list
end

defp selected?(option, value) do
  html_escape(option) == html_escape(value)
end
```

This is only a rough example adapted from a project. The basic idea is to end the checkbox name with `[]`, which Phoenix will parse as a list. The form would produce something like:

```elixir
%{
  "filters" => %{
    "0" => %{
      "field" => "status_filter",
      "op" => "in",
      "value" => ["succeeded", "failed"]
    }
  }
}
```

**Comment by egze** [comment]
Managed to get the desired behaviour with `multiple_select`, but still puzzled if it's cleanly possible with checkboxes.

**Edit:** didn't see that you answered above. Will study it, thanks.
